File: games/game_tigerdragon/DragonTigerConfig/Lua/DragonTiger_CardsCFG.lua
DragonTiger_CardsCFG = {}

function DragonTiger_CardsCFG:getData(key)
	if self.datas == nil then
		return nil
	end
	return self.datas[key]
end

function DragonTiger_CardsCFG:init()
	self.datas = {}
	self.datas[0] = {CardsID = 0, CardsName = "没牛", CardsRate = 1, CardsTypeStr = "@A", SoundID = 2, CardsAnimation = ""}
	self.datas[1] = {CardsID = 1, CardsName = "牛1", CardsRate = 1, CardsTypeStr = "01", SoundID = 3, CardsAnimation = ""}
	self.datas[2] = {CardsID = 2, CardsName = "牛2", CardsRate = 1, CardsTypeStr = "02", SoundID = 4, CardsAnimation = ""}
	self.datas[3] = {CardsID = 3, CardsName = "牛3", CardsRate = 1, CardsTypeStr = "03", SoundID = 5, CardsAnimation = ""}
	self.datas[4] = {CardsID = 4, CardsName = "牛4", CardsRate = 1, CardsTypeStr = "04", SoundID = 6, CardsAnimation = ""}
	self.datas[5] = {CardsID = 5, CardsName = "牛5", CardsRate = 1, CardsTypeStr = "05", SoundID = 7, CardsAnimation = ""}
	self.datas[6] = {CardsID = 6, CardsName = "牛6", CardsRate = 1, CardsTypeStr = "06", SoundID = 8, CardsAnimation = ""}
	self.datas[7] = {CardsID = 7, CardsName = "牛7", CardsRate = 2, CardsTypeStr = "07", SoundID = 9, CardsAnimation = ""}
	self.datas[8] = {CardsID = 8, CardsName = "牛8", CardsRate = 3, CardsTypeStr = "08", SoundID = 10, CardsAnimation = ""}
	self.datas[9] = {CardsID = 9, CardsName = "牛9", CardsRate = 4, CardsTypeStr = "09", SoundID = 11, CardsAnimation = ""}
	self.datas[10] = {CardsID = 10, CardsName = "牛牛", CardsRate = 5, CardsTypeStr = "00", SoundID = 12, CardsAnimation = ""}
	self.datas[11] = {CardsID = 11, CardsName = "银牛", CardsRate = 6, CardsTypeStr = ";0", SoundID = 13, CardsAnimation = "NIUNIUyinniudonghua"}
	self.datas[12] = {CardsID = 12, CardsName = "四炸", CardsRate = 7, CardsTypeStr = "<=", SoundID = 14, CardsAnimation = "niuniubaozhatexiao"}
	self.datas[13] = {CardsID = 13, CardsName = "五花牛", CardsRate = 8, CardsTypeStr = ">B0", SoundID = 15, CardsAnimation = "NIUNIUjinniudonghua"}
	self.datas[14] = {CardsID = 14, CardsName = "五小牛", CardsRate = 9, CardsTypeStr = ">?0", SoundID = 16, CardsAnimation = "NIUNIUwuxiaoniudonghua"}
end

DragonTiger_CardsCFG:init()

File: games/game_tigerdragon/DragonTigerConfig/Code/Client/DragonTiger_CardsCFG.cpp
#include "stdafx.h"
#include "tinyxml2/tinyxml2.h"
#include "LuaCfgHelper.h"
#include <iostream>
#include <boost/algorithm/string.hpp>
#include "DragonTiger_CardsCFG.h"
std::auto_ptr<DragonTiger_CardsCFG> DragonTiger_CardsCFG::msSingleton(nullptr);

int DragonTiger_CardsCFG::GetCount()
{
	return (int)mMapData.size();
}

const DragonTiger_CardsCFGData* DragonTiger_CardsCFG::GetData(int CardsID)
{
	auto it = mMapData.find(CardsID);
	if (it != mMapData.end())
	{
		return &it->second;
	}
	return NULL;
}

const std::map<int, DragonTiger_CardsCFGData>& DragonTiger_CardsCFG::GetMapData()
{
	return mMapData;
}

void DragonTiger_CardsCFG::Load()
{
	tinyxml2::XMLDocument xmlDoc;
	std::string content = FileUtils::getInstance()->getStringFromFile("Config/DragonTiger_CardsCFG.xml");
	auto result = xmlDoc.Parse(content.c_str(), content.length());
	if (result != tinyxml2::XML_SUCCESS)
	{
		CCLOGERROR("Result:%d", result);
		CCASSERT(false, "result != tinyxml2::XML_SUCCESS");
		return;
	}
	auto root = xmlDoc.RootElement();
	if (root == NULL)
	{
		CCASSERT(false, "root == NULL");
		return;
	}
	auto element = root->FirstChildElement("Data");
	while (element != NULL)
	{
		DragonTiger_CardsCFGData data;
		data.mCardsID = element->IntAttribute("CardsID");
		data.mCardsName = element->Attribute("CardsName");
		data.mCardsRate = element->IntAttribute("CardsRate");
		data.mCardsTypeStr = element->Attribute("CardsTypeStr");
		data.mSoundID = element->IntAttribute("SoundID");
		data.mCardsAnimation = element->Attribute("CardsAnimation");
		if (mMapData.find(data.mCardsID) != mMapData.end())std::cout <<"data refind:" << data.mCardsID << std::endl;
		CCASSERT(mMapData.find(data.mCardsID) == mMapData.end(), "data.mCardsID is exists");
		mMapData.insert(std::make_pair(data.mCardsID, data));
		element = element->NextSiblingElement();
	}
	CCLOG("DragonTiger_CardsCFG Loaded. Load Data:%u", mMapData.size());
}

void DragonTiger_CardsCFG::LoadLua()
{
	LuaEngine::getInstance()->executeScriptFile("Config/DragonTiger_CardsCFG");
	lua_State* L = LuaEngine::getInstance()->getLuaStack()->getLuaState();
	lua_getglobal(L, "DragonTiger_CardsCFG");
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushstring(L, "datas");
	lua_gettable(L, -2);
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushnil(L);
	while(lua_next(L, 2))
	{
		CCASSERT(lua_istable(L, -1) == 1, "is not table");
		DragonTiger_CardsCFGData data;
		LuaCfgHelper::readInt(L, "CardsID", data.mCardsID);
		LuaCfgHelper::readString(L, "CardsName", data.mCardsName);
		LuaCfgHelper::readInt(L, "CardsRate", data.mCardsRate);
		LuaCfgHelper::readString(L, "CardsTypeStr", data.mCardsTypeStr);
		LuaCfgHelper::readInt(L, "SoundID", data.mSoundID);
		LuaCfgHelper::readString(L, "CardsAnimation", data.mCardsAnimation);
		if (mMapData.find(data.mCardsID) != mMapData.end())std::cout <<"data refind:" << data.mCardsID << std::endl;
		CCASSERT(mMapData.find(data.mCardsID) == mMapData.end(), "data.mCardsID is exists");
		mMapData.insert(std::make_pair(data.mCardsID, data));
		lua_pop(L, 1);
	}
	lua_settop(L, 0);
	CCLOG("DragonTiger_CardsCFG Loaded. Load Data:%u", mMapData.size());
}

void DragonTiger_CardsCFG::Reload()
{
	mMapData.clear();
	Load();
}

DragonTiger_CardsCFG* DragonTiger_CardsCFG::GetSingleton()
{
	if (msSingleton.get() == nullptr)
	{
		msSingleton.reset(new DragonTiger_CardsCFG());
	}
	return msSingleton.get();
}
